Jormun: geovelo's call to direct path

pbougue opened the pull request:
TODO:

- [x] Warn when direct_path != isochrone (one more call to API, maybe in debug) ? 
> Not so usefull for now

- [x] Attach "correctly" fallback to tc (experimental, in another PR) ? 
> Done in this PR (hand-tested with geovelo, a bigger refacto is needed on experimental)

Commit:
--clang warning

Commit:
refactor valhalla

extract polyline decode (to be used by geovelo)
improve section management

Commit:
geovelo: call direct path

Commit:
test a bit easier to read

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -52,16 +52,33 @@ def __init__(self, instance, service_url, id='geovelo', timeout=10, api_key=None
+                {'latitude': coord_dest.lat, 'longitude': coord_dest.lon}
+            ],
+            'bikeDetails': {
+                'profile': 'BEGINNER',
```
open question, are there different interesting profiles ? shouldn't this be a connector configuration ?

pbougue replied:
I add a comment on possible values.
It seems too soon to put that in the connector or at call (following profile and bike speed?). In my opinion, we have to wait until geovelo manages parameters consistently in isochrone and in direct_path.

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -125,69 +142,98 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
                                    .format(len(origins), len(destinations)))
 
-        data = self._make_request_arguments(origins, destinations)
+        data = Geovelo._make_request_arguments_isochrone(origins, destinations)
```
you can use `cls` here

xlqian replied:
data = ~~Geovelo~~cls._make_....

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -125,69 +142,98 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
         r = self._call_geovelo('{}/{}'.format(self.service_url, 'api/v2/routes_m2m'),
                                requests.post, json.dumps(data))
-        self._check_response(r)
+        Geovelo._check_response(r)
```
same here

xlqian replied:
~~Geovelo~~cls._check_....

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -125,69 +142,98 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
             raise UnableToParse('Geovelo nb response != nb requested')
 
-        return self._get_matrix(resp_json)
+        return Geovelo._get_matrix(resp_json)
```
you can use `cls` here

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -196,14 +242,20 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, datetime, c
 
-        data = Geovelo._make_request_arguments([pt_object_origin], [pt_object_destination])
-        r = self._call_geovelo('{}/{}'.format(self.service_url, 'api/v2/routes_m2m'),
+        data = Geovelo._make_request_arguments_direct_path(pt_object_origin, pt_object_destination)
```
you can use `self._make...`

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -196,14 +242,20 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, datetime, c
+                                                                'objects_as_ids=true&'),
                                requests.post, json.dumps(data))
-        self._check_response(r)
+        Geovelo._check_response(r)
```
you can use `self`

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -196,14 +242,20 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, datetime, c
             raise UnableToParse('Geovelo nb response != nb requested')
 
-        return self._get_response(resp_json, mode, pt_object_origin, pt_object_destination, datetime, clockwise)
+        return Geovelo._get_response(resp_json, mode, pt_object_origin, pt_object_destination, datetime, clockwise)
```
you can use `self`

antoine-de commented on source/jormungandr/jormungandr/street_network/valhalla.py:
```diff
@@ -140,36 +112,40 @@ def _get_response(self, json_resp, mode, pt_object_origin, pt_object_destination
-            else:
-                journey.departure_date_time = datetime - journey.duration
-                journey.arrival_date_time = datetime
+        journey = resp.journeys.add()
```
I don't understand why you changed this ?

pbougue replied:
We used to add one journey per leg (which is a section).
It was invisible as we are supposed to have only one leg, but it was not consistent. I did the cleanup along the comparison while I tried to use pretty much the same thing in geovelo.

xlqian replied:
@antoine-de he pulled out some one-shot code from the loop

antoine-de commented on source/jormungandr/jormungandr/street_network/valhalla.py:
```diff
@@ -201,10 +178,10 @@ def _get_valhalla_mode(self, kraken_mode):
 
-        valhalla_mode = self._get_valhalla_mode(mode)
-        destinations = [self._format_coord(destination, api) for destination in pt_object_destinations]
+        valhalla_mode = Valhalla._get_valhalla_mode(mode)
```
you can keep `self` there

Commit:
github remarks

Commit:
little update on contributing

Commit:
stick fallback to pt (even if not consistent with isochrone)

antoine-de commented on source/jormungandr/jormungandr/street_network/geovelo.py:
```diff
@@ -52,16 +52,33 @@ def __init__(self, instance, service_url, id='geovelo', timeout=10, api_key=None
+            'bikeDetails': {
+                'profile': 'BEGINNER', # can be MEDIAN, EXPERT
+                'bikeType': 'TRADITIONAL', # can be 'BSS'
+                # 'averageSpeed': '6' # in km/h, BEGINER sets it to 13
```
shouldn't we use the query parameters there ?

TeXitoi replied:
when we can do it on the matrix call, but not before

pbougue replied:
To elaborate a bit: yes we should, but there would be even more difference between isochrone and directpath.
As geovelo's parameters may change when they will implem speed for isochrone, and also as we may change a bit experimental when refactoring it (even though this part shouldn't change much)... It's does not seem worth doing it now.

Commit:
many little refacto (naming, corrections)

among renaming :
reverse_sections -> is_fallback_at_end
(in fallback context) clockwise -> is_start_fallback_datetime
fallback_direct_path -> fallback_direct_path_pool
no more "_from" and "to"

Commit:
really stick fallback to pt (and a little refacto)

antoine-de commented on source/jormungandr/jormungandr/scenarios/experimental.py:
```diff
@@ -78,46 +78,57 @@ def get_max_fallback_duration(request, mode, dp_durations={}):
+def make_direct_path_key(dep_mode, orig_uri, dest_uri, is_fallback_at_end,
+                         datetime, is_start_fallback_datetime):
+    '''
+    A, B and mode matters obviously
```
A and B does not appear above, it's a bit confusing

and maybe add a bit more general comment first

antoine-de commented on source/jormungandr/jormungandr/scenarios/experimental.py:
```diff
@@ -391,56 +402,52 @@ def get_update_crowfly_duration_futures(self):
-    def _get_direct_path(instance, mode, pt_object_origin, pt_object_destination, datetime, clockwise,
-                         request, reverse_sections=False):
+    def _get_direct_path(instance, mode, pt_object_origin, pt_object_destination,
+                         datetime, is_start_fallback_datetime, request, is_fallback_at_end):
```
what is is_start_fallback_datetime ?

xlqian commented on source/jormungandr/jormungandr/scenarios/experimental.py:
```diff
@@ -42,11 +42,11 @@
 import copy
 
-def create_crowfly(pt_journey, _from, to, begin, end, mode='walking'):
+def create_crowfly(pt_journey, crowfly_origin, crowfly_destination, begin, end, mode='walking'):
```
👍

xlqian commented on source/jormungandr/jormungandr/scenarios/experimental.py:
```diff
@@ -466,63 +473,63 @@ def worker_journey(dep_mode=dep_mode,
+        pt_journey.durations.walking += fallback_copy.journeys[0].durations.walking
+        _extend_pt_sections_with_direct_path(pt_journey, fallback_copy)
+
+    def _build_from(self, journey, first_section, journey_origin, crowfly_stop_points, odt_stop_points,
```
so we may rename `journey` to `pt_journey`?

Commit:
github remarks : refactor for clarity
